server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala

```diff
@@ -351,6 +348,21 @@ class InteractiveSession(
   _appId = appIdHint
   sessionStore.save(RECOVERY_SESSION_TYPE, recoveryMetadata)
 
+  // TODO Replace this with a Rpc call from repl to server.
+  private val stateThread = new Thread(new Runnable {
+    override def run(): Unit = {
+      try {
+        while (_state.isActive) {
+          // State is also updated when we get statement results from repl, not just here.
+          setSessionStateFromReplState(client.map(_.getReplState.get()))
+          Thread.sleep(1000)
+        }
+      } catch {
+        case _: InterruptedException =>
+      }
+    }
+  })
+
```

Comment:
I agree with @zjffdu , this will create one thread per session, also do we need to pull the repl state so frequently? Each session will send RPC requests per second, that's a big overhead if there's lots of session.